### PR TITLE
Add support for LCS Command

### LINKED ISSUE #2403: Add support for LCS command
Redis LCS command was added in version 7.0.0 and is currently supported by the library. We should add support as documented [here](https://redis.io/commands/LCS).

### PR DESCRIPTION
**Pull request by SoulPancake**
Fixes #2403

### CODE REVIEW

File: command.go [hunk 1]
```diff
@@ -3840,3 +3840,54 @@ func (cmd *ZSliceWithKeyCmd) readReply(rd *proto.Reader) (err error) {
...
+type LCSMatch struct {
+	Positions1 []int
+	Positions2 []int
+}
+
+type LCSMatchLen struct {
+	Match LCSMatch
+	Len int64
+}
+
```
Comment (SoulPancake) [inline command.go:3855]: @monkey92t @vmihailenco  How does this look like for the command response for LCS with Len and Can you suggest a good structure for MINMATCHLEN and  WITHMATCHLEN?
Should I try to create a common structure? Any suggestions for that

**Comment by monkey92t** [comment]
my thoughts:

```go
type LCSMatch struct {
	Matches []LCSMatchedPosition
	Len     int
}

type LCSMatchedPosition struct {
        // or A/B ?
	Key1     LCSPosition
	Key2     LCSPosition

        // only for withMatchLen is true
	MatchLen int
}

type LCSPosition struct {
	Start int
	End   int
}

// lcs key1 key2
func LCS(key1, key2 string) *StringCmd {}

// lcs key1 key2 len
func LCSLen(key1, key2 string) *IntCmd {}

// lcs key1 key2 idx minmatchlen minMatchLen withmatchlen
// or *LCSMatchCmd
func LCSIdx(key1, key2 string, minMatchLen int, withMatchLen bool) *LCSCmd {}
```

😀😀😀😀

**Comment by SoulPancake** [comment]
> my thoughts:
> 
> ```go
> type LCSMatch struct {
> 	Matches []LCSMatchedPosition
> 	Len     int
> }
> 
> type LCSMatchedPosition struct {
>         // or A/B ?
> 	Key1     LCSPosition
> 	Key2     LCSPosition
> 
>         // only for withMatchLen is true
> 	MatchLen int
> }
> 
> type LCSPosition struct {
> 	Start int
> 	End   int
> }
> 
> // lcs key1 key2
> func LCS(key1, key2 string) *StringCmd {}
> 
> // lcs key1 key2 len
> func LCSLen(key1, key2 string) *IntCmd {}
> 
> // lcs key1 key2 idx minmatchlen minMatchLen withmatchlen
> // or *LCSMatchCmd
> func LCSIdx(key1, key2 string, minMatchLen int, withMatchLen bool) *LCSCmd {}
> 
> //--------------
> 
> // OR
> type LCSQuery struct {
> 	Key1         string
> 	Key2         string
> 	Len          bool
> 	Idx          bool
> 	MinMatchLen  int
> 	WithMatchLen bool
> }
> 
> func LCS(key1, key2 string, len, idx bool, minMatchLen int, withMatchLen bool) *LCSCmd {}
> func LCS(q *LCSQuery) *LCSCmd{}
> ```
> 
> 😀😀😀😀 comments are welcome~

If we take a common LCSQuery as the params then also we have to separately handle the return types right?
My doubt is whether we have to do the third as LCSMatchCmd or two/three different LCSWithIDX LCSwithMinmatchlength etc.

**Comment by monkey92t** [comment]
exp:

```go
type LCSMatch struct {
	MatchString string
	Matches     []LCSMatchedPosition
	Len         int64
}

type LCSMatchedPosition struct {
	// or A/B ?
	Key1 LCSPosition
	Key2 LCSPosition

	// only for withMatchLen is true
	MatchLen int
}

type LCSPosition struct {
	Start int
	End   int
}

type LCSCmd struct {
	baseCmd

	// 1: match string
	// 2: match len
	// 3: match idx LCSMatch
	readType uint8
	val      *LCSMatch
}

func (cmd *LCSCmd) readReply(rd *proto.Reader) (err error) {
	lcs := &LCSMatch{}
	switch cmd.readType {
	case 1:
		// match string
		lcs.MatchString, err = rd.ReadString()
	case 2:
		// match len
		if lcs.Len, err = rd.ReadInt(); err != nil {
			return err
		}
	case 3:
		// read LCSMatch
		if err = rd.ReadFixedMapLen(2); err != nil {
			return err
		}

		key, err := rd.ReadString()
		if err != nil {
			return err
		}

		// we should not depend on the order of the map.
		switch key {
		case "matches":
			n, err := rd.ReadArrayLen()
			if err != nil && err != Nil {
				return err
			}
			for i := 0; i < n; i++ {
				// read LCSMatchedPosition
				// miss..
			}
		case "len":
			if lcs.Len, err = rd.ReadInt(); err != nil {
				return err
			}
		}
	}

	cmd.val = lcs
	return nil
}

type LCSQuery struct {
	Key1         string
	Key2         string
	Len          bool
	Idx          bool
	MinMatchLen  int
	WithMatchLen bool
}

func LCS(ctx context.Context, q *LCSQuery) *LCSCmd {
	// ...miss
	return &LCSCmd{}
}

// call
lcs, err := rdb.LCS(ctx, &LCSQuery{
	//...
}).Result()

lcs.MatchString()
lcs.Len()
lcs.Matches[0].Key1.Start
```

**Comment by monkey92t** [comment]
any thoughts on my example?

**Comment by SoulPancake** [comment]
> any thoughts on my example?

@monkey92t I am trying to build using your example only as you can see in the latest pushes

**Comment by SoulPancake** [comment]
Hi @monkey92t 
Need some help here
I am not sure if I should implement the MarshalBinary() function for the interface slice or should I read some other way
I am having difficulty in this, Can you please guide me

**Comment by monkey92t** [comment]
I don't quite understand what you mean. What does this have to do with MarshalBinary?

**Comment by monkey92t** [comment]
No, that's not it. I will make some modifications to the PR later.

**Comment by SoulPancake** [comment]
Thanks @monkey92t :) I was overcomplicating it

**Comment by SoulPancake** [comment]
@monkey92t  As this looks good , we can merge it right?
Is there anything else left to do here?

**Review by SoulPancake — COMMENTED** [review]
LGTM

File: commands_test.go [hunk 2]
```diff
@@ -2251,6 +2251,75 @@ var _ = Describe("Commands", func() {
...
+					Key1:     redis.LCSPosition{Start: 2, End: 3},
+					Key2:     redis.LCSPosition{Start: 0, End: 1},
+					MatchLen: 0,
+				},
+			}))
+
+			lcs, err = client.LCS(ctx, &redis.LCSQuery{
+				Key1:         "key1",
+				Key2:         "key2",
+				Idx:          true,
```
[the remark this answers is not in the document]
Comment (monkey92t) [reply commands_test.go:2307]: Yes, it is possible to respond with errors. The lcs command has multiple situations where errors can occur: 
1. key is not a string type
2. command parameter error
3. data is too large
4. insufficient memory
...
We just need to verify that the LCS API can handle errors normally without worrying about the specific error types. 
Are you interested? You can try to add some tests.

[the remark this answers is not in the document]
Comment (SoulPancake) [reply commands_test.go:2307]: @monkey92t As the struct itself will validate the key type and command params so I guess that case is redundant as we're not writing validators in general

How do you suggest we test the data being too large?
Something like the following?
```
largeString := strings.Repeat("x", 11000000)
err := client.MSet(ctx, "key1", "oh"+largeString, "key2", "my"+largeString).Err()
Expect(err).To(HaveOccurred())
lcs, err := client.LCS(ctx, &redis.LCSQuery{
    Key1: "key1",
    Key2: "key2",
}).Result()
Expect(err).To(HaveOccurred())

```

[the remark this answers is not in the document]
Comment (monkey92t) [reply commands_test.go:2307]: We don't need to test for memory overflow, and such tests are very unstable because we can't predict the size of the memory. We only need to test for incorrect data types to verify if the API handles errors correctly.

exp:
```shell
> set key1 "mytest"
> lpush key2 one
> lcs key1 key2
```

[the remark this answers is not in the document]
Comment (monkey92t) [reply commands_test.go:2307]: `key2` is not of string type, and the lcs command cannot process it, so it will return an error.

[the remark this answers is not in the document]
Comment (SoulPancake) [reply commands_test.go:2307]: Understood, will add these tests

File: commands_test.go [hunk 3]
```diff
@@ -2251,6 +2251,90 @@ var _ = Describe("Commands", func() {
...
+					Key1:     redis.LCSPosition{Start: 4, End: 7},
+					Key2:     redis.LCSPosition{Start: 5, End: 8},
+					MatchLen: 4,
+				},
+			}))
+
+			_, err = client.Set(ctx, "keywithstringvalue", "golang",0).Result()
+			Expect(err).NotTo(HaveOccurred())
+           
+            _, err = client.LPush(ctx, "keywithnonstringvalue", "somevalue").Result()
```
Comment (monkey92t) [inline commands_test.go:2325]: gofmt

File: commands_test.go [hunk 4]
```diff
@@ -2251,6 +2251,90 @@ var _ = Describe("Commands", func() {
...
+			Expect(err).NotTo(HaveOccurred())
+
+			_, err = client.LCS(ctx, &redis.LCSQuery{
+				Key1: "keywithstringvalue",
+				Key2: "keywithnonstringvalue",
+			}).Result()
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(Equal("ERR The specified keys must contain string values"))
+
+
```
Comment (monkey92t) [inline commands_test.go:2335]: remove excessive blank lines...

File: commands_test.go (hunk 3 above)
Comment (SoulPancake) [reply commands_test.go]: resolved

File: commands_test.go (hunk 4 above)
Comment (SoulPancake) [reply commands_test.go]: Done

File: commands_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (SoulPancake) [reply commands_test.go:2307]: @monkey92t Skipping the following error as it will be avoided by using readType in args
image

[the remark this answers is not in the document]
Comment (monkey92t) [reply commands_test.go:2307]: Yes, we just need to verify if go-redis can handle errors without focusing on specific error types. Redis-server may increase or decrease its error responses, and we are not writing unit tests for redis-server.

**Comment by SoulPancake** [comment]
@monkey92t The build failure seems unrelated

**Comment by monkey92t** [comment]
> @monkey92t The build failure seems unrelated

We use redis-server/sentinel/cluster to perform tests, and sometimes encounter failures due to network errors.

**Comment by SoulPancake** [comment]
@monkey92t Had a general question, how were you able to run the build actions again without pushing a new commit?

**Review by monkey92t — APPROVED** [review]

**Comment by monkey92t** [comment]
> @monkey92t Had a general question, how were you able to run the build actions again without pushing a new commit?

You need to manually build or rebuild the projects that you have permissions for in Github Actions.
